test/unit/cmap/connection.test.ts

```diff
@@ -0,0 +1,248 @@
+
+      const beforeHandshakeSymbol = getSymbolFrom(err, 'beforeHandshake');
+      expect(err).to.have.property(beforeHandshakeSymbol, true);
+
+      done();
+    });
+  });
+
+  describe('onTimeout()', () => {
+    let connection: sinon.SinonSpiedInstance<Connection>;
+    let clock: sinon.SinonFakeTimers;
+    let driverSocket: sinon.SinonSpiedInstance<FakeSocket>;
+    let messageStream: MessageStream;
+    let kDelayedTimeoutId: symbol;
+    let NodeJSTimeoutClass: any;
+
+    /** The absolute minimum socket API needed by Connection as of writing this test */
+    class FakeSocket extends EventEmitter {
+      address() {
+        // is never called
+      }
+      pipe() {
+        // does not need to do anything
+      }
+      destroy() {
+        // is called, has no side effects
+      }
+      get remoteAddress() {
+        return 'iLoveJavaScript';
+      }
+      get remotePort() {
+        return 123;
+      }
+    }
+
+    beforeEach(() => {
+      clock = sinon.useFakeTimers();
+
+      NodeJSTimeoutClass = setTimeout(() => null, 1).constructor;
+
+      driverSocket = sinon.spy(new FakeSocket());
+      // @ts-expect-error: This subset of options is all that is needed
+      connection = sinon.spy(new Connection(driverSocket, { id: 1 }));
+      const messageStreamSymbol = getSymbolFrom(connection, 'messageStream');
+      kDelayedTimeoutId = getSymbolFrom(connection, 'delayedTimeoutId');
+      messageStream = connection[messageStreamSymbol];
+    });
+
+    afterEach(() => {
+      clock.restore();
+    });
+
+    it('should delay timeout errors by one tick', async () => {
+      expect(connection).to.have.property(kDelayedTimeoutId, null);
+
+      driverSocket.emit('timeout');
+      expect(connection.onTimeout).to.have.been.calledOnce;
+      expect(connection).to.have.property(kDelayedTimeoutId).that.is.instanceOf(NodeJSTimeoutClass);
+
+      clock.tick(1);
+
+      expect(driverSocket.destroy).to.have.been.calledOnce;
+      expect(connection).to.have.property('closed', true);
+      // timeout callback should clear it's own reference
+      expect(connection).to.have.property(kDelayedTimeoutId, null);
+    });
+
+    it('should clear timeout errors if more data is available', () => {
+      expect(connection).to.have.property(kDelayedTimeoutId, null);
+
+      driverSocket.emit('timeout');
+      expect(connection.onTimeout).to.have.been.calledOnce;
+      expect(connection).to.have.property(kDelayedTimeoutId).that.is.instanceOf(NodeJSTimeoutClass);
+
+      // emit a message before the clock ticks even once
+      // onMessage ignores unknown 'responseTo' value
+      messageStream.emit('message', { responseTo: null });
+
+      // New message before clock ticks 1 will clear the timeout
+      expect(connection).to.have.property(kDelayedTimeoutId, null);
+
+      // ticking the clock should do nothing, there is no timeout anymore
+      clock.tick(1);
+
+      expect(driverSocket.destroy).to.not.have.been.calledOnce;
```

Comment:
the correct assertion here is to `not.have.been.called` (at all), because not being called once doesn't rule out have been called more than once